### PR TITLE
Refactor update ongoing analyses

### PR DESCRIPTION
**Pull request by henrikstranneheim**
## Description

Refactor update ongoing analyses

### Added

- Tests

### Changed

- Refactor update ongoing analyses and migrate to CRUD

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b refactor_update_ongoing_analyses -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: tests/store/crud/test_read.py [hunk 1]
```diff
@@ -90,6 +90,19 @@ def test_get_analyses_for_case(analysis_store: MockStore, case_id: str):
     assert analyses
 
 
+def test_get_analyses_with_statuses(analysis_store: MockStore):
+    """Test getting analyses with status when it exists in the database."""
+    # GIVEN a store with an analysis
+
+    # WHEN accessing it by statuses
```
Comment (ChrOertlin) [inline tests/store/crud/test_read.py:97]:
```suggestion
    # WHEN accessing an analysis by statuses
```

File: tests/store/crud/test_read.py [hunk 2]
```diff
@@ -90,6 +90,19 @@ def test_get_analyses_for_case(analysis_store: MockStore, case_id: str):
...
+    analyses: Optional[List[Analysis]] = analysis_store.get_analyses_with_statuses(
+        statuses=list(TrailblazerStatus.ongoing_statuses())
+    )
+
+    # THEN it should return analyses
```
Comment (ChrOertlin) [inline tests/store/crud/test_read.py:102]:
```suggestion
    # THEN an analysis should be returned
```

File: tests/store/crud/test_read.py [hunk 3]
```diff
@@ -90,6 +90,19 @@ def test_get_analyses_for_case(analysis_store: MockStore, case_id: str):
...
+    assert analyses
```
Comment (ChrOertlin) [inline tests/store/crud/test_read.py:103]:
```suggestion
    assert analyses
```
We should add an assertion that the returned analysis has the status it was queried by

File: tests/store/crud/test_update.py [hunk 4]
```diff
@@ -50,6 +50,57 @@ def test_update_user_is_archived(user_store: MockStore, user_email: str):
...
+    # GIVEN an analysis store and analysis status
+    analysis: Optional[Analysis] = analysis_store.get_latest_analysis_for_case(
+        case_id=ongoing_analysis_case_id
+    )
+    analysis.status = TrailblazerStatus.COMPLETED
+
+    # WHEN updating ongoing analyses
+    analysis_store.update_ongoing_analyses()
+
+    # THEN status should be the same
```
Comment (ChrOertlin) [inline tests/store/crud/test_update.py:81]:
```suggestion
    # THEN status should be COMPLETED
```

File: tests/store/crud/test_update.py [hunk 5]
```diff
@@ -50,6 +50,57 @@ def test_update_user_is_archived(user_store: MockStore, user_email: str):
...
+
+def test_update_ongoing_analyses_with_not_ongoing_analysis(
+    analysis_store: MockStore, ongoing_analysis_case_id: str
+):
+    """Test all ongoing analysis statuses are updated on not ongoing analysis-"""
+    # GIVEN an analysis store and analysis status
+    analysis: Optional[Analysis] = analysis_store.get_latest_analysis_for_case(
+        case_id=ongoing_analysis_case_id
+    )
+    analysis.status = TrailblazerStatus.COMPLETED
```
Comment (ChrOertlin) [inline tests/store/crud/test_update.py:76]:
```suggestion
    # GIVEN a completed analysis
    analysis.status = TrailblazerStatus.COMPLETED
```

File: tests/store/crud/test_update.py [hunk 6]
```diff
@@ -50,6 +50,57 @@ def test_update_user_is_archived(user_store: MockStore, user_email: str):
...
+def test_update_ongoing_analyseswhen_bad_call(
+    analysis_store: MockStore, caplog, ongoing_analysis_case_id: str
+):
+    """Test all ongoing analysis statuses are updated when exception is raised."""
+    caplog.set_level("INFO")
+    # GIVEN an analysis store and analysis status
+    analysis: Optional[Analysis] = analysis_store.get_latest_analysis_for_case(
+        case_id=ongoing_analysis_case_id
+    )
+    del analysis.id
```
Comment (ChrOertlin) [inline tests/store/crud/test_update.py:94]:
```suggestion
# GIVEN an analysis that is deleted
    del analysis.id
```

File: trailblazer/store/crud/update.py [hunk 7]
```diff
@@ -32,6 +32,19 @@ def update_user_is_archived(self, user: User, archive: bool = True) -> None:
...
 
+    def update_ongoing_analyses(self, analysis_host: Optional[str] = None) -> None:
+        """Iterate over all analysis with ongoing status and query SLURM for current progress."""
+        ongoing_analyses: Optional[List[Analysis]] = self.get_analyses_with_statuses(
+            statuses=list(TrailblazerStatus.ongoing_statuses())
+        )
+        for analysis in ongoing_analyses:
+            try:
+                self.update_run_status(analysis_id=analysis.id, analysis_host=analysis_host)
+            except Exception as error:
```
Comment (ChrOertlin) [inline trailblazer/store/crud/update.py:43]:
```suggestion
            except Exception as error:
```
Can we make the scope of this exception smaller?

**Review by ChrOertlin — APPROVED** [review]
Looks good!

File: trailblazer/store/crud/update.py (hunk 7 above)
Comment (henrikstranneheim) [reply trailblazer/store/crud/update.py:44]: Would be good - but out of scope

File: tests/store/crud/test_read.py (hunk 3 above)
Comment (henrikstranneheim) [reply tests/store/crud/test_read.py:103]: Added

**Comment by henrikstranneheim** [comment]
### CLI and API ✅ 

```
ilblazer, version 18.0.13
[henrik.stranneheim@hasta:~] [S_base] 1m45s $ trailblazer scan
Error updating analysis for: case - grandaphid : PermissionError
Error updating analysis for: case - goodbengal : MissingSqueueOutput
Error updating analysis for: case - helpedlab : MissingSqueueOutput
Error updating analysis for: case - newgopher : PermissionError
Error updating analysis for: case - onestag : PermissionError
Error updating analysis for: case - funnyaphid : FileNotFoundError
Error updating analysis for: case - proudfish : PermissionError
Error updating analysis for: case - setcolt : PermissionError
Error updating analysis for: case - handyshrimp : FileNotFoundError
Error updating analysis for: case - divinetomcat : PermissionError
Error updating analysis for: case - HW2V2DRX2 : FileNotFoundError
Error updating analysis for: case - savingstag : FileNotFoundError
Error updating analysis for: case - topewe : PermissionError
Error updating analysis for: case - fullgoat : PermissionError
Error updating analysis for: case - smoothtuna : FileNotFoundError
Error updating analysis for: case - cosmicslug : PermissionError
Error updating analysis for: case - readyhorse : PermissionError
Error updating analysis for: case - valueddog : MissingSqueueOutput
Error updating analysis for: case - lovingkoala : PermissionError
Error updating analysis for: case - topmarmot : PermissionError
Error updating analysis for: case - rarecow : PermissionError
Error updating analysis for: case - maturebeagle : PermissionError
Error updating analysis for: case - fithusky : PermissionError
Error updating analysis for: case - moralsnake : PermissionError
Error updating analysis for: case - drivenearwig : PermissionError
Error updating analysis for: case - nativetick : PermissionError
Error updating analysis for: case - smoothworm : PermissionError
Error updating analysis for: case - honestmoray : PermissionError
Error updating analysis for: case - poeticimpala : PermissionError
Error updating analysis for: case - mainsalmon : PermissionError
Error updating analysis for: case - knownloon : PermissionError
Error updating analysis for: case - merrybat : PermissionError
Error updating analysis for: case - quietdoe : PermissionError
Error updating analysis for: case - goldencolt : PermissionError
Error updating analysis for: case - aliveswine : PermissionError
Error updating analysis for: case - bigdragon : PermissionError
Error updating analysis for: case - fitburro : PermissionError
Error updating analysis for: case - uniquegrub : PermissionError
Error updating analysis for: case - fullnewt : PermissionError
Error updating analysis for: case - coolswine : PermissionError
Error updating analysis for: case - fleethyena : PermissionError
Error updating analysis for: case - smoothwombat : PermissionError
Error updating analysis for: case - coolcougar : PermissionError
Error updating analysis for: case - busytoad : PermissionError
Error updating analysis for: case - knownbull : PermissionError
Error updating analysis for: case - clearliger : PermissionError
Error updating analysis for: case - sunnybaboon : PermissionError
Error updating analysis for: case - flyingimp : PermissionError
Error updating analysis for: case - stableimp : PermissionError
Error updating analysis for: case - goodrobin : PermissionError
Error updating analysis for: case - comicmite : PermissionError
Error updating analysis for: case - rapidnewt : PermissionError
Error updating analysis for: case - valuedlab : PermissionError
Error updating analysis for: case - secureghoul : PermissionError
Error updating analysis for: case - hipgull : PermissionError
Error updating analysis for: case - livingmoth : PermissionError
Error updating analysis for: case - puredrum : PermissionError
Error updating analysis for: case - grandkite : PermissionError
Error updating analysis for: case - soundminnow : PermissionError
Error updating analysis for: case - cosmicsponge : PermissionError
Error updating analysis for: case - holydory : PermissionError
Error updating analysis for: case - solidpoodle : PermissionError
Error updating analysis for: case - warmghoul : PermissionError
Error updating analysis for: case - fondbengal : PermissionError
Error updating analysis for: case - awaregull : PermissionError
Error updating analysis for: case - onlocust : PermissionError
Error updating analysis for: case - modelrhino : PermissionError
Error updating analysis for: case - intentalien : PermissionError
Error updating analysis for: case - notedbat : PermissionError
Error updating analysis for: case - nobleflea : PermissionError
Error updating analysis for: case - soughtcub : PermissionError
Error updating analysis for: case - stablekite : PermissionError
Error updating analysis for: case - holyfowl : PermissionError
Error updating analysis for: case - starmacaw : PermissionError
Error updating analysis for: case - toughcoral : PermissionError
Error updating analysis for: case - neededhawk : PermissionError
Error updating analysis for: case - causaldonkey : PermissionError
Error updating analysis for: case - movingcivet : PermissionError
Error updating analysis for: case - nextwalrus : PermissionError
Error updating analysis for: case - warmcivet : PermissionError
Error updating analysis for: case - steadypup : PermissionError
Error updating analysis for: case - divinesnake : PermissionError
Error updating analysis for: case - bossmidge : PermissionError
Error updating analysis for: case - savedsnipe : PermissionError
Error updating analysis for: case - oncorgi : PermissionError
Error updating analysis for: case - largeliger : PermissionError
Error updating analysis for: case - alertox : PermissionError
Error updating analysis for: case - suitedelk : PermissionError
Error updating analysis for: case - hardycivet : PermissionError
Error updating analysis for: case - epicfalcon : PermissionError
Error updating analysis for: case - growntapir : PermissionError
Error updating analysis for: case - rightmonkey : PermissionError
Error updating analysis for: case - rightwhale : PermissionError
Error updating analysis for: case - valuedgibbon : PermissionError
Error updating analysis for: case - nativegannet : PermissionError
Error updating analysis for: case - richwombat : PermissionError
Error updating analysis for: case - firmcoral : PermissionError
Error updating analysis for: case - grandhare : PermissionError
Error updating analysis for: case - wiredkodiak : PermissionError
Error updating analysis for: case - simplerat : PermissionError
Error updating analysis for: case - brightmayfly : PermissionError
Error updating analysis for: case - pumpedhyena : PermissionError
Error updating analysis for: case - tidyant : PermissionError
Error updating analysis for: case - firstcod : PermissionError
Error updating analysis for: case - mutualjoey : PermissionError
Error updating analysis for: case - tidyshrew : PermissionError
Error updating analysis for: case - strongprawn : PermissionError
Error updating analysis for: case - holycicada : PermissionError
Error updating analysis for: case - restedmako : PermissionError
Error updating analysis for: case - normalfish : PermissionError
Error updating analysis for: case - busywren : PermissionError
Error updating analysis for: case - fleetturtle : PermissionError
Error updating analysis for: case - freshoriole : PermissionError
Error updating analysis for: case - directamoeba : PermissionError
Error updating analysis for: case - gameoyster : PermissionError
Error updating analysis for: case - setracer : PermissionError
Error updating analysis for: case - gentlemolly : PermissionError
Error updating analysis for: case - vastswift : PermissionError
Error updating analysis for: case - upwardperch : PermissionError
Error updating analysis for: case - divinewolf : PermissionError
Error updating analysis for: case - quickcougar : PermissionError
Error updating analysis for: case - actualkid : PermissionError
Error updating analysis for: case - grandnewt : PermissionError
Error updating analysis for: case - caringmullet : PermissionError
Error updating analysis for: case - sacredpoodle : PermissionError
Error updating analysis for: case - funnylouse : PermissionError
Error updating analysis for: case - sharplark : PermissionError
Error updating analysis for: case - mastercattle : PermissionError
Error updating analysis for: case - freshhound : PermissionError
Error updating analysis for: case - quietmutt : PermissionError
Error updating analysis for: case - guidedmutt : PermissionError
Error updating analysis for: case - fithawk : PermissionError
Error updating analysis for: case - modelhornet : PermissionError
Error updating analysis for: case - wholekodiak : PermissionError
Error updating analysis for: case - kindlab : PermissionError
Error updating analysis for: case - freegator : PermissionError
Error updating analysis for: case - modesteft : PermissionError
Error updating analysis for: case - proudwalrus : PermissionError
Error updating analysis for: case - minttroll : PermissionError
Error updating analysis for: case - manycobra : PermissionError
Error updating analysis for: case - gamesawfly : PermissionError
Error updating analysis for: case - holysatyr : PermissionError
Error updating analysis for: case - politecobra : PermissionError
Error updating analysis for: case - livingtahr : PermissionError
Error updating analysis for: case - solidclam : PermissionError
Error updating analysis for: case - cleverbaboon : PermissionError
Error updating analysis for: case - propercat : PermissionError
Error updating analysis for: case - modesthare : PermissionError
Error updating analysis for: case - primepigeon : PermissionError
Error updating analysis for: case - gladshark : PermissionError
Error updating analysis for: case - suitedpiglet : PermissionError
Error updating analysis for: case - eagerzebra : PermissionError
Error updating analysis for: case - divinewalrus : FileNotFoundError
Updated status civilsole - 279436: failed
All analyses updated!
```

Screenshot 2023-10-12 at 09 05 02

**Comment by henrikstranneheim** [comment]
Deployed to production:

```
Log deploy... done.
trailblazer, version 18.0.14
```

**Pull request merged**